sys/compat/linprocfs/linprocfs.c

```diff
@@ -2092,6 +2095,176 @@ linprocfs_domax_map_cnt(PFS_FILL_ARGS)
 	return (0);
 }
 
+/*
+ * Filler function for proc/sysvipc/msg
+ */
+static int
+linprocfs_dosysvipc_msg(PFS_FILL_ARGS)
+{
+	struct msqid_kernel *msqids;
+	u_long id, msgmni;
+	size_t sz;
+	int error;
+
+	sbuf_printf(sb,
+	    "%10s %10s %4s  %10s %10s %5s %5s %5s %5s %5s %5s %10s %10s %10s\n",
+	    "key", "msqid", "perms", "cbytes", "qnum", "lspid", "lrpid",
+	    "uid", "gid", "cuid", "cgid", "stime", "rtime", "ctime");
+
+again:
+	msgmni = msginfo.msgmni;
+	sz = sizeof(struct msqid_kernel) * msgmni;
+	msqids = malloc(sz, M_TEMP, M_NOWAIT);
+	if (msqids == NULL)
+		return (ENOMEM);
+	if (msgmni != msginfo.msgmni) {
+		free(msqids, M_TEMP);
+		goto again;
+	}
+
+	error = kernel_sysctlbyname(curthread, "kern.ipc.msqids", msqids, &sz,
+	    NULL, 0, 0, 0);
+	if (error != 0) {
+		free(msqids, M_TEMP);
+		return (error);
+	}
+	msgmni = sz / sizeof(struct msqid_kernel);
+
+	for (id = 0; id < msgmni; id++)
+		if (msqids[id].u.msg_qbytes != 0)
+			sbuf_printf(sb,
+			    "%10d %10lu  %4o  %10lu %10lu %5u %5u %5u %5u %5u %5u %10ld %10ld %10ld\n",
+			    (int) msqids[id].u.msg_perm.key,
+			    IXSEQ_TO_IPCID(id, msqids[id].u.msg_perm),
+			    msqids[id].u.msg_perm.mode,
+			    msqids[id].u.msg_cbytes,
+			    msqids[id].u.msg_qnum,
+			    msqids[id].u.msg_lspid,
+			    msqids[id].u.msg_lrpid,
+			    msqids[id].u.msg_perm.uid,
+			    msqids[id].u.msg_perm.gid,
+			    msqids[id].u.msg_perm.cuid,
+			    msqids[id].u.msg_perm.cgid,
+			    msqids[id].u.msg_stime,
+			    msqids[id].u.msg_rtime,
+			    msqids[id].u.msg_ctime);
+
+	free(msqids, M_TEMP);
+	return (0);
+}
+
+/*
+ * Filler function for proc/sysvipc/sem
+ */
+static int
+linprocfs_dosysvipc_sem(PFS_FILL_ARGS)
+{
+	struct semid_kernel *semids;
+	u_long id, semmni;
+	size_t sz;
+	int error;
+
+	sbuf_printf(sb, "%10s %10s %4s %10s %5s %5s %5s %5s %10s %10s\n",
+	    "key", "semid", "perms", "nsems", "uid", "gid", "cuid", "cgid",
+	    "otime", "ctime");
+
+again:
+	semmni = seminfo.semmni;
+	sz = sizeof(struct semid_kernel) * semmni;
+	semids = malloc(sz, M_TEMP, M_NOWAIT);
+	if (semids == NULL)
+		return (ENOMEM);
+	if (semmni != seminfo.semmni) {
+		free(semids, M_TEMP);
+		goto again;
+	}
+
```

Comment:
Same as above. I think removing this check and keying off ENOMEM in the sysctl return value is the cleaner way to deal with this since we don't seem to have a cleaner, race-free interface to extract this info.

Reply:
> Same as above. I think removing this check and keying off ENOMEM in the sysctl return value is the cleaner way to deal with this since we don't seem to have a cleaner, race-free interface to extract this info.

What do you mean by "keying off ENOMEM" ?